fix: Load TEI in TrackerCapture app trigger update query

vietnguyen opened the pull request:
https://jira.dhis2.org/browse/DHIS2-9230

- Removed `AbstractDeviceRenderTypeMapDeserializer`
- Add function `programAttributeStore.getAttributes( programs )`  and  `entityTypeAttributeStore.getAttributes( trackedEntityTypes)`
Those new functions extract TrackedEntityAttribute  directly from hibernate, no mapping/deserialization needed. It also uses one `IN` condition instead of n queries by `programid` as in old code. 

This fix considered as a work around, it doesn't fix the root cause. Maybe the cause is because of jsonb column deserialization causing additional update queries.  Need more time to investigate and work on that.

luciano-fiandesio commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramTrackedEntityAttributeStore.java:
```diff
@@ -39,4 +42,6 @@
 {
     ProgramTrackedEntityAttribute get( Program program, TrackedEntityAttribute attribute );
+
+    List<TrackedEntityAttribute> getAttributesByProgramsWithDataRead( List<Program> programs, User user );
```
1) Please add Javadoc
2) I would rename this method to `getAttributes` and explain the ACL part in the java docs

vietnguyen replied:
Fixed and added unit test

luciano-fiandesio commented on dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityTypeAttributeStore.java:
```diff
@@ -28,13 +28,14 @@
-    {
-        super( SectionRenderingObject::new );
-    }
+    List<TrackedEntityAttribute> getAttributesByTrackedEntityTypesWithDataRead( List<TrackedEntityType> trackedEntityTypes, User user );
```
1) Please add Javadoc
  2) I would rename this method to getAttributes and explain the ACL part in the java docs

vietnguyen replied:
Fixed and added unit test

luciano-fiandesio commented on dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/jsonb/type/JsonDeviceRenderTypeMap.java:
```diff
@@ -28,38 +28,32 @@
-    protected Object convertJsonToObject( String content )
+    protected JavaType getResultingJavaType( Class<?> returnedClass )
     {
-        try
```
I'm not very familiar with this renderer, but this look like a drastic change. Can you comment what has been done here?

vietnguyen replied:
In an attempt to fix the issue, I have removed AbstractDeviceRenderTypeMapDeserializer which I have added to support XML deserialization of RenderType. So the content of the method `JsonDeviceRenderTypeMap.convertMapToObject()` is actually copied from the deserializer. Even if this doesn't solve the issue, I think it reduces the complexity of the code.
This is covered by the test `MetadataImportServiceTest.testSerializeDeviceRenderTypeMap()`

luciano-fiandesio commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramTrackedEntityAttributeStore.java:
```diff
@@ -66,4 +75,21 @@ public ProgramTrackedEntityAttribute get(Program program, TrackedEntityAttribute
     }
+
+    @Override
+    public List<TrackedEntityAttribute> getAttributesByProgramsWithDataRead( List<Program> programs, User user )
```
Do we have a test for this function?

vietnguyen replied:
Added unit test

vietnguyen commented:
> Test coverage for this PR is quite low (10%), do you think we can improve it?

@luciano-fiandesio I added unit test. Also moved the data read check out of the DAO functions to the `DefaultTrackedEntityAttributeService. getAllUserReadableTrackedEntityAttributes()` function. I think it should be faster that way.

stian-sandvold commented on dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackedEntityAttributeService.java:
```diff
@@ -283,13 +288,20 @@ else if ( trackedEntityAttribute.hasOptionSet() && !trackedEntityAttribute.isVal
+        {
+            attributes.addAll( entityTypeAttributeStore.getAttributes(
+                trackedEntityTypes.stream().filter( trackedEntityType -> aclService.canDataRead( user, trackedEntityType ) ).collect(
+                    Collectors.toList()) ));
```
```suggestion
                    Collectors.toList() ) ) );
```
